src/transformers/models/ibert/modeling_ibert.py

```diff
@@ -0,0 +1,1418 @@
+        return position_ids.unsqueeze(0).expand(input_shape)
+
+
+# Referred from transformers.models.bert.modeling_bert.BertSelfAttention
+class IBertSelfAttention(nn.Module):
+    def __init__(self, config):
+        super().__init__()
+        if config.hidden_size % config.num_attention_heads != 0 and not hasattr(config, "embedding_size"):
+            raise ValueError(
+                "The hidden size (%d) is not a multiple of the number of attention "
+                "heads (%d)" % (config.hidden_size, config.num_attention_heads)
+            )
+        self.quant_mode = config.quant_mode
+        self.weight_bit = 8
+        self.bias_bit = 32
+        self.act_bit = 8
+
+        self.num_attention_heads = config.num_attention_heads
+        self.attention_head_size = int(config.hidden_size / config.num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attention_head_size
+
+        # Q, K, V Linear layers
+        self.query = QuantLinear(self.weight_bit, bias_bit=self.bias_bit, 
+                quant_mode=self.quant_mode, per_channel=True)
+        self.key = QuantLinear(self.weight_bit, bias_bit=self.bias_bit, 
+                quant_mode=self.quant_mode, per_channel=True)
+        self.value = QuantLinear(self.weight_bit, bias_bit=self.bias_bit, 
+                quant_mode=self.quant_mode, per_channel=True)
+        self.query.set_param(nn.Linear(config.hidden_size, self.all_head_size))
+        self.key.set_param(nn.Linear(config.hidden_size, self.all_head_size))
+        self.value.set_param(nn.Linear(config.hidden_size, self.all_head_size))
+
+        # Requantization (32bit -> 8bit) for Q, K, V activations
+        self.query_activation = QuantAct(self.act_bit, quant_mode=self.quant_mode)
+        self.key_activation = QuantAct(self.act_bit, quant_mode=self.quant_mode)
+        self.value_activation = QuantAct(self.act_bit, quant_mode=self.quant_mode)
+        self.output_activation = QuantAct(self.act_bit, quant_mode=self.quant_mode)
+
+        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
+        self.position_embedding_type = getattr(config, "position_embedding_type", "absolute")
+        assert self.position_embedding_type == "absolute"
+        self.is_decoder = config.is_decoder
+        assert not self.is_decoder
+
+        self.softmax = IntSoftmax(self.act_bit, quant_mode=self.quant_mode)
+
+    def transpose_for_scores(self, x):
+        new_x_shape = x.size()[:-1] + (self.num_attention_heads, self.attention_head_size)
+        x = x.view(*new_x_shape)
+        return x.permute(0, 2, 1, 3)
+
+    def forward(
+        self,
+        hidden_states,
+        hidden_states_scaling_factor,
+        attention_mask=None,
+        head_mask=None,
+        encoder_hidden_states=None,
+        encoder_attention_mask=None,
+        past_key_value=None,
+        output_attentions=False,
+    ):
+        assert encoder_hidden_states is None
+        assert past_key_value is None
```

Comment:
Could we add messages to the assertions so that we know what's the issue if it crashes?

Reply:
I would just fully delete the functionality if it's not supported tbh

Reply:
Removed them